core/src/test/scala/unit/kafka/server/OffsetFetchRequestTest.scala

```diff
@@ -14,233 +14,532 @@
-        val topicData = response.data().topics().get(0)
-        val partitionData = topicData.partitions().get(0)
-        if (version < 3) {
-          assertEquals(AbstractResponse.DEFAULT_THROTTLE_TIME, response.throttleTimeMs())
-        }
-        verifySingleGroupResponse(version.asInstanceOf[Short],
-          response.error().code(), partitionData.errorCode(), topicData.name(),
-          partitionData.partitionIndex(), partitionData.committedOffset(),
-          partitionData.committedLeaderEpoch(), partitionData.metadata())
+  @ClusterTest(serverProperties = Array(
+    new ClusterConfigProperty(key = "unstable.api.versions.enable", value = "true"),
+    new ClusterConfigProperty(key = "group.coordinator.new.enable", value = "true"),
+    new ClusterConfigProperty(key = "group.consumer.max.session.timeout.ms", value = "600000"),
+    new ClusterConfigProperty(key = "group.consumer.session.timeout.ms", value = "600000"),
+    new ClusterConfigProperty(key = "offsets.topic.num.partitions", value = "1"),
+    new ClusterConfigProperty(key = "offsets.topic.replication.factor", value = "1")
+  ))
+  def testMultiGroupsOffsetFetchWithNewConsumerGroupProtocolAndNewGroupCoordinator(): Unit = {
+    testMultipleGroupsOffsetFetch(useNewProtocol = true, requireStable = true)
+  }
+
+  @ClusterTest(serverProperties = Array(
+    new ClusterConfigProperty(key = "unstable.api.versions.enable", value = "true"),
+    new ClusterConfigProperty(key = "group.coordinator.new.enable", value = "true"),
+    new ClusterConfigProperty(key = "group.consumer.max.session.timeout.ms", value = "600000"),
+    new ClusterConfigProperty(key = "group.consumer.session.timeout.ms", value = "600000"),
+    new ClusterConfigProperty(key = "offsets.topic.num.partitions", value = "1"),
+    new ClusterConfigProperty(key = "offsets.topic.replication.factor", value = "1")
+  ))
+  def testMultiGroupsOffsetFetchWithOldConsumerGroupProtocolAndNewGroupCoordinator(): Unit = {
+    testMultipleGroupsOffsetFetch(useNewProtocol = false, requireStable = false)
+  }
+
+  @ClusterTest(clusterType = Type.ALL, serverProperties = Array(
+    new ClusterConfigProperty(key = "unstable.api.versions.enable", value = "false"),
+    new ClusterConfigProperty(key = "group.coordinator.new.enable", value = "false"),
+    new ClusterConfigProperty(key = "group.consumer.max.session.timeout.ms", value = "600000"),
+    new ClusterConfigProperty(key = "group.consumer.session.timeout.ms", value = "600000"),
+    new ClusterConfigProperty(key = "offsets.topic.num.partitions", value = "1"),
+    new ClusterConfigProperty(key = "offsets.topic.replication.factor", value = "1")
+  ))
+  def testMultiGroupsOffsetFetchWithOldConsumerGroupProtocolAndOldGroupCoordinator(): Unit = {
+    testMultipleGroupsOffsetFetch(useNewProtocol = false, requireStable = true)
+  }
+
+  private def testSingleGroupOffsetFetch(useNewProtocol: Boolean, requireStable: Boolean): Unit = {
```

Comment:
This method is a little large and contains quite a few smaller test cases within. Is there a way we can break it up a bit more?

Reply:
Sure. I can separate the fetch offsets vs fetch all offsets cases.